apps/zotonic_mod_cron/src/mod_cron.erl

```diff
@@ -37,48 +52,41 @@
     terminate/2
 ]).
 
--export([ tick_event/2 ]).
-
 -record(state, {
-          site :: atom(),
-          tick_job_refs = [] :: list(erlcron:job_ref())
+          jobs = #{},
+          context :: z:context()
          }).
 
-%% Default timers. These are called approximately every N seconds.
--define(TIMER_INTERVAL, [
-    {{ 1, sec}, tick_1s},
-    {{ 1, min}, tick_1m},
-    {{10, min}, tick_10m},
-    {{15, min}, tick_15m},
-    {{30, min}, tick_30m},
-    {{ 1, hr},  tick_1h},
-    {{ 2, hr},  tick_2h},
-    {{ 3, hr},  tick_3h},
-    {{ 4, hr},  tick_4h},
-    {{ 6, hr},  tick_6h},
-    {{12, hr},  tick_12h},
-    {{24, hr},  tick_24h}
-]).
-
 -include_lib("zotonic_core/include/zotonic.hrl").
 
-tick_event(Event, Site) ->
-    z_notifier:notify(Event, z_context:new(Site)).
-
 start_link(Args) when is_list(Args) ->
+    gen_server:start_link(?MODULE, Args, []).
+
+init(Args) ->
+    erlang:process_flag(trap_exit, true),
     {context, Context} = proplists:lookup(context, Args),
+
     Site = z_context:site(Context),
-    gen_server:start_link(?MODULE, Site, []).
+    logger:set_process_metadata(#{
+        site => Site,
+        module => ?MODULE
+    }),
 
-init(Site) when is_atom(Site) ->
-    erlang:process_flag(trap_exit, true),
-    z_context:logger_md(Site),
-    JobRefs = start_tick_jobs(Site),
-    {ok, #state{ site = Site, tick_job_refs = JobRefs }}.
+    z_notifier:observe(module_activate, self(), Context),
+    z_notifier:observe(module_deactivate, self(), Context),
+
+    State = add_jobs(?MODULE, #state{context = Context}),
+    {ok, State}.
 
 handle_call(_Msg, _From, State) ->
     {reply, {error, unknown_message}, State}.
 
+handle_cast({#module_activate{ module=Module }, _Context}, State) ->
+        State1 = add_jobs(Module, State),
+    {noreply, State1};
+handle_cast({#module_deactivate{ module=Module }, _Context}, State) ->
+    State1 = cancel_jobs(Module, State),
```

Comment:
the API `cancel_jobs/2` always return atom `ok`. I can be wrong but should be used real state at `{noreply, State}`. But specifically for this clause keep atom `ok` as state is satisfied.